### ISSUE TITLE
kill "Press ENTER" with fire

### ISSUE THREAD

**Issue by jwgarber** [neovim/neovim#1029]
I'm not very familiar with C or else I'd do it myself, but it would be great if the 'Press ENTER or type command to continue' line didn't print every time you ran an external command. It's something you only need to see once, and seeing it every time gets rather annoying. I think this message should just be removed and the behaviour it describes (hitting a key to go back to vim after running a shell command) be folded into the documentation (if it's not there already).

**Comment by arames** [comment]
As a user of :! to execute commands, I agree with the conclusions of Shugyousha and DanTheColoradan. I use the pause and prompt to execute commands for commands with a text output, but I would welcome an option to not have the pause and prompt. Opening another terminal is not always convenient (eg. work over ssh or using a small screen).

**Comment by alexdunn** [comment]
I would LOVE to see the abolition of the Press Enter to continue nonsense.  At the very least, an option to disable it.  

Let's say for example you have UltiSnips installed, which has an autocmd triggered on every cursor move, and your Python installation is broken, so on every cursor move UtliSnips will give an error.  How utterly insane.  vim would be perfectly usable (just without UltiSnips) with a broken Python.  It shouldn't become completely unusable just because a single plugin isn't happy.

**Comment by gaving** [comment]
@alexdunn :+1:

**Comment by arames** [comment]
@alexdunn 
In the issue you mention, is the plugin actually using ':!' to execute commands, or is it only showing a similar message and prompt given by another mechanism? Isn't the issue more that the plugin uses (or has to use?) an unadapted mechanism to execute commands? Or that it doesn't handle errors properly?

As a user facing feature the prompt is not nonsense for everyone. Some people do use this feature, so I don't think it should be completely removed. An option to configure this behaviour seems like a good idea. I personally don't mind about the default value for the option. I think it depends on the neovim policy chosen for the new defaults. Backward compatibility would suggest keeping the default behaviour similar to vim; user friendly defaults would lean toward skipping the prompt all the time or when nothing was output.

Personally I'd be fine with a 'user friendly' default for this. 'Advanced' users will know to look for ":help" when something doesn't work as they expect after switching to neovim. As for plugins, they should have a way to handle errors properly without having to modify the behaviour of user facing commands.

**Comment by justinmk** [comment]
I have a PR in progress that help with this.

**Comment by noscript** [comment]
So annoying to have that message for shell commands. When echoing two lines `:echo "line 1" | echo "line 2" ` with `cmdheight=2` the message is not displayed. How on earth it is enforced for shell commands, even if the output can perfectly fit into the command line.

> I think there's an improvement that can be made without compromising any existing use cases: if a command returns empty, don't show a "Press ENTER" prompt; instead, just echo a "shell returned " status message.

Couldn't agree more.

**Comment by substack** [comment]
Can this `Press ENTER` message and pause be removed from the source tree everywhere so that I never, ever see it? It shows up in many more places than just shell commands.

**Comment by justinmk** [comment]
Adding an option to avoid it should be feasible. I'm sure you could predict that removing it totally is not an option.

**Comment by tarruda** [comment]
This message has always bugged me too, but the only way to get rid of it is to completely change how nvim presents notifications to the user. One solution I had considered is to use a special `:terminal` buffer that will receive every message from the editor, which would allow messages(command output and error) to be presented to the user without blocking the UI with the press ENTER state. More details are here: https://github.com/neovim/neovim/pull/2076#issuecomment-76595194

**Comment by azure-satellite** [comment]
> I think there's an improvement that can be made without compromising any existing use cases: if a command returns empty, don't show a "Press ENTER" prompt; instead, just echo a "shell returned " status message.
I think this is very reasonable

**Comment by wsdjeg** [comment]
I also get same issue with 
```viml
func! s:test()
   set nomore
   echo "sss\nsss"
endf
au VimEnter * call timer_start(400, function('s:test'))
```

**Comment by fred-drake** [comment]
For anybody who finds this in a web search, I got around this by adding an additional `<CR>` at the end of my command.  It's not exactly what is being asked for, but for many it should be a suitable workaround.

**Comment by kylefhartzenberg** [comment]
This has been an issue for me when attempting to use `vim.api.nvim_echo()` to display information that usually ends up being wider than the window size. Setting `cmdheight` prior to calling `vim.api.nvim_echo()` with the intent to reset it post-echo didn't seem to help. Based on the information in https://github.com/neovim/neovim/issues/16462, I wrote the following function that truncates the middle section of the message in order to display a message that has some start/end context and also doesn't elicit the forsaken 'Press ENTER or type command to continue'.
```lua
test_message = "This long message will not display on a single line on most screens. If it does, ensure your window size is smaller."

local function non_disruptive_echo(message)
    trunc_fill = "..."
    first_part = string.sub(message, 1, vim.v.echospace/2 - #trunc_fill)
    second_part = string.sub(message, #message - math.floor((vim.v.echospace/2)), #message)
    vim.api.nvim_echo({{first_part .. trunc_fill .. second_part}}, false, {})
end

non_disruptive_echo(test_message)
```
Output on a very small window: `This long messag...dow size is smaller.`

It would be really good to have an option to disable the 'Press ENTER...' prompt when using `vim.api.nvim_echo()` and still display the full string to be echoed.

**Comment by Shougo** [comment]
The option can be added.  But I don't know it is better solution.

**Comment by Shougo** [comment]
@kylefhartzenberg You can use `nomore` option for it.

**Comment by kylefhartzenberg** [comment]
@Shougo Could you please provide an example of how to use the `nomore` option? The documentation for [nvim_echo](https://neovim.io/doc/user/api.html#nvim_echo()) doesn't seem to mention anything about it so I am not sure what you have in mind.

**Comment by Shougo** [comment]
Please see the documentation.

```
						*'more'* *'nomore'*
'more'			boolean	(default: on)
			global
	When on, listings pause when the whole screen is filled.  You will get
	the |more-prompt|.  When this option is off there are no pauses, the
	listing continues until finished.
```

**Comment by Shougo** [comment]
Hm...  more does not work for `press-enter` prompt.

But you can use it.

```vim
function! s:echo(msg) abort
  const more_save = &more
  const showcmd_save = &showcmd
  const ruler_save = &ruler
  try
    set nomore
    set noshowcmd
    set noruler

    const height = [1, &cmdheight]->max()
    echo ''
    for i in range(0, a:msg->len() - 1, height)
      redraw

      let m = a:msg[i : i+height-1]->join("\n")
      echo m
      if has('vim_starting')
        echo ''
      endif
    endfor
  finally
    let &more = more_save
    let &showcmd = showcmd_save
    let &ruler = ruler_save
  endtry
endfunction
```

**Comment by kylefhartzenberg** [comment]
@Shougo Yes, setting `nomore` doesn't seem to help when it comes to `nvim_echo`. I've tried to convert your function to the equivalent Lua code along with the appropriate Neovim API calls but there are still some issues (I don't have much experience with VimL to Lua conversion). Could you please have a look and provide the Lua equivalent so I can test whether this function provides the desired behaviour?
```lua
local function echo(message)
    -- Save original values
    local more_orig = vim.opt.more:get()
    local showcmd_orig = vim.opt.showcmd:get()
    local ruler_orig = vim.opt.ruler:get()

    -- Set new temporary values
    vim.opt.more = false
    vim.opt.showcmd = false
    vim.opt.ruler = false
 
    local height = math.max(1, vim.opt.cmdheight:get())
    
    vim.api.nvim_echo({{""}}, false, {})

    local m
    for i = 1, #message, height do -- note Lua uses 1 indexing
        vim.fn.redraw()
        m = string.sub(message, i, i+height-1) -- Add join("\n") equivalent here?
        vim.api.nvim_echo({{m}}, false, {})
        -- Add has('has_starting') equivalent here?
    end

    -- Reset original values
    vim.opt.more = more_orig
    vim.opt.showcmd = showcmd_orig
    vim.opt.ruler = ruler_orig
end
```

**Comment by Shougo** [comment]
Hi.  I don't know about the Lua...

This script works for me.

```vim
function! s:echo(msgs) abort
  const more_save = &more
  const showcmd_save = &showcmd
  const ruler_save = &ruler
  try
    set nomore
    set noshowcmd
    set noruler

    const height = [1, &cmdheight]->max()
    echo ''
    for i in range(0, a:msgs->len() - 1, height)
      redraw

      let m = a:msgs[i : i+height-1]->join("\n")
      echo m
    endfor
  finally
    let &more = more_save
    let &showcmd = showcmd_save
    let &ruler = ruler_save
  endtry
endfunction
set cmdheight=3
call s:echo("foo\nbar\nbaz"->split("\n"))
```

**Comment by Shougo** [comment]
` vim.fn.redraw()` must be `vim.cmd.redraw()`.

`m = string.sub(message, i, i+height-1)` does not work.

```lua
local function echo(msgs)
    -- Save original values
    local more_orig = vim.opt.more:get()
    local showcmd_orig = vim.opt.showcmd:get()
    local ruler_orig = vim.opt.ruler:get()

    -- Set new temporary values
    vim.opt.more = false
    vim.opt.showcmd = false
    vim.opt.ruler = false
 
    local height = math.max(1, vim.opt.cmdheight:get())
    
    vim.api.nvim_echo({{""}}, false, {})

    local m
    for i = 1, #msgs, height do -- note Lua uses 1 indexing
        vim.cmd.redraw()
        -- NOTE: does not work
        --m = string.sub(message, i, i+height-1) -- Add join("\n") equivalent here?
        --vim.api.nvim_echo({{m}}, false, {})
        -- Add has('has_starting') equivalent here?
    end

    -- Reset original values
    vim.opt.more = more_orig
    vim.opt.showcmd = showcmd_orig
    vim.opt.ruler = ruler_orig
end
echo({"foo", "bar", "baz"})
```

**Comment by Shougo** [comment]
You can use `messagesopt` option for it.

https://github.com/neovim/neovim/commit/ec94c2704f5059794923777ed51412d80bd26b5b

**Comment by justinmk** [comment]
Thanks for working on 'messagesopt' @Shougo .

This still isn't a general solution though; an obvious problem with 'messagesopt' is that the user still must decide to either (1) get interrupted by a "sleep", or (2) hide all output, even output from intentional commands such as `:messages`. 

# Next steps

There are some improvements we should make regardless of 'messagesopt':  

- [ref](https://github.com/neovim/neovim/issues/1029#issuecomment-52362039) : 
  > if a command returns empty, don't show a "Press ENTER" prompt; instead, just echo a "shell returned " status message.
- https://github.com/neovim/neovim/issues/22478
- and specifically https://github.com/neovim/neovim/issues/22478#issuecomment-2480239231 : 
  > Always truncate messages
  - should "always truncate" be a flag on 'messagesopt' ?

**Comment by zeertzjq** [comment]
Maybe it can truncate the message if neither `hit-enter` nor `wait:` is present?

**Comment by Shougo** [comment]
> This still isn't a general solution though; an obvious problem with 'messagesopt' is that the user still must decide to either (1) get interrupted by a "sleep", or (2) hide all output, even output from intentional commands such as :messages.

Yes.  I know it is not perfect solution.  But it works and it is better than previous.
I use `messagesopt` toggle for `:messages`.

> Maybe it can truncate the message if neither hit-enter nor wait: is present?

I considered implementing truncate support for messagesopt, but I couldn't see any advantages of enabling truncate, so I decided not to implement.
I think it only makes sense when `wait:` is `0`, but if nothing is displayed on the screen, it doesn't matter whether truncate is enabled or not.

**Comment by justinmk** [comment]
Closed by https://github.com/neovim/neovim/pull/27855

It's currently opt-in. Try it like this:

    require('vim._extui').enable({})

- shows a [+n] hint (instead of "Press ENTER") if messages exceed 'cmdheight' (use "g<" to see all)
- messages pager is now a real buffer+window
- ":" cmdline has syntax highlighting
